python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from typing import Any, DefaultDict, Tuple, Type, TypeVar
+from weakref import WeakSet
+
+import rmm
+
+from cudf.core.buffer.buffer import Buffer, cuda_array_interface_wrapper
+
+T = TypeVar("T", bound="CopyOnWriteBuffer")
+
+
+class CopyOnWriteBuffer(Buffer):
+    """A Buffer represents device memory.
+
+    Use the factory function `as_buffer` to create a Buffer instance.
+    """
+
+    # This dict keeps track of all instances that have the same `ptr`
+    # and `size` attributes.  Each key of the dict is a `(ptr, size)`
+    # tuple and the corresponding value is a set of weak references to
+    # instances with that `ptr` and `size`.
+    _instances: DefaultDict[Tuple, WeakSet] = defaultdict(WeakSet)
```

Comment:
This is a (slow) memory leak, I think, because when the final (ptr, size) pair for a buffer disappears it is not evicted from this dictionary.

We could either implement `__del__` on this class to cleanup the `_instances` dict when things eventually go out of scope, or make this a WeakKeyDictionary (needs a little bit of playing around):

```
class PtrSize:
    __slots__ = ("t", "__weakref__")
    def __init__(self, ptr, size):
        self.t = (ptr, size)
    def __hash__(self):
        return hash(self.t)
    def __eq__(self, o):
         return type(o) is type(self) and o.t == self.t
    def __repr__(self):
         return f"PtrSize{self.t!r}"
    def __str__(self):
        return f"PtrSize{self.t}"


class CoWBuffer:
    _instances = WeakKeyDictionary()
    ...
    def __init__(self, ...):
        self.ptrsize = PtrSize(ptr, size)
    
    def _finalize_init(self):
        self.__class__._instances.setdefault(self.ptrsize, WeakSet()).add(self)
        ...
```
Now, assuming that ptrsize objects are only internal, as soon as the last buffer referencing one goes away, the key disappears from the dictionary.

Reply:
I think using a `WeakKeyDictionary` here with `PtrAndSize` as keys makes sense

Reply:
Switched to using `WeakKeyDictionary` & `PtrAndSize`(made it internal with `_PtrAndSize`)

Reply:
Closing the loop on this one, after discussing offline with @wence- we decided to revert usage of `WeakKeyDictionary` because the entries in the dict are being discarded if the strong reference to first key is lost:
```python
In [1]: import cudf

In [2]: cudf.set_option("copy_on_write", True)

In [3]: s1 = cudf.Series([1, 2, 3, 4])
s
In [4]: s2 = s1.copy(deep=False)

In [5]: s3 = s2.copy(deep=False)

In [6]: cudf.core.buffer.CopyOnWriteBuffer._instances[s1._column.base_data._PtrAndSize]
Out[6]: {<weakref at 0x7f295e32a360; to 'CopyOnWriteBuffer' at 0x7f2a0594da00>, <weakref at 0x7f2a04e82e00; to 'CopyOnWriteBuffer' at 0x7f2a058dbbe0>, <weakref at 0x7f2a04e89c70; to 'CopyOnWriteBuffer' at 0x7f2a040e2df0>}

In [7]: del s1

In [8]: del s2

In [9]: cudf.core.buffer.CopyOnWriteBuffer._instances[s3._column.base_data._PtrAndSize]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[9], line 1
----> 1 cudf.core.buffer.CopyOnWriteBuffer._instances[s3._column.base_data._PtrAndSize]

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/weakref.py:416, in WeakKeyDictionary.__getitem__(self, key)
    415 def __getitem__(self, key):
--> 416     return self.data[ref(key)]

KeyError: <weakref at 0x7f295e2d24a0; to '_PtrAndSize' at 0x7f296e6477f0>


In [11]: cudf.core.buffer.CopyOnWriteBuffer._instances.keyrefs()
Out[11]: []
```

Reply:
Ah I see, the `_keys_cleanup` finalizer below handles this now, is that right?

Reply:
> Ah I see, the `_keys_cleanup` finalizer below handles this now, is that right?

Yup, that's correct.